src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,536 @@
+import { CssClassesOnly, DraggableSlide, Transition } from './transitions';
+import { mathf, Raf } from '../..';
+import { EventDispatcher, EventManager } from '../events';
+import { CarouselSynchronizer } from './carousel-synchronizer';
+
+const DEFAULT_DISTANCE_TO_ACTIVE_SLIDE_ATTR = 'data-index';
+
+/**
+ * For readability, specifies which half should be gotten by providing a
+ * direction to loop through elements.
+ */
+enum Direction {
+  LEFT = -1,
+  RIGHT = 1
+}
+
+/**
+ * Carousel events.
+ */
+export enum CarouselEvent {
+  TRANSITION_START = 'transitionStart',
+  TRANSITION_END = 'transitionEnd'
+}
+
+/**
+ * Options provided to the constructor
+ */
+export interface CarouselOptions {
+  activeCssClass?: string;
+  beforeCssClass?: string;
+  afterCssClass?: string;
+  allowLooping?: boolean;
+  distanceToActiveSlideAttr?: string;
+  // Transition used to iterate through slides.
+  transition?: Transition|string;
+  // Carousel code won't run if the condition is not met.
+  // Used for establishing multiple types of carousels on the same DOM varied
+  // by breakpoint.
+  condition?: () => boolean;
+  // Time to stay on a slide before automatically transitioning to the next
+  // slide
+  autoplaySpeed?: number;
+}
+
+/**
+ * Default classes applied to slide elements.
+ */
+enum DefaultCssClass {
+  ACTIVE_SLIDE = 'active',
+  BEFORE_SLIDE = 'before',
+  AFTER_SLIDE = 'after'
+}
+
+/**
+ * Tracks an element that should be transitioned to.
+ *
+ * Groups the two pieces of information that are needed to transition a carousel
+ * to the given element.
+ *
+ * @param element The element that needs to be transitioned to.
+ * @param drivenBySync Whether this transition is driven by a carousel sync.
+ *                     If it is driven by sync, then we know that we don't need
+ *                     to bother the CarouselSynchronizer for another sync.
+ */
+class TransitionTarget {
+  readonly element: HTMLElement;
+  readonly drivenBySync: boolean;
+
+  constructor(element: HTMLElement, drivenBySync: boolean = false) {
+    this.element = element;
+    this.drivenBySync = drivenBySync;
+  }
+}
+
+export class Carousel implements EventDispatcher {
+  autoplaySpeed: number;
+  private readonly activeCssClass: string;
+  private readonly beforeCssClass: string;
+  private readonly afterCssClass: string;
+  private readonly distanceToActiveSlideAttr: string;
+  private readonly condition: () => boolean;
+  private readonly container: HTMLElement;
+  private readonly slides: HTMLElement[];
+  private readonly transition: Transition;
+  private readonly allowLooping: boolean;
```

Comment:
Can we call this `loop` instead?

Reply:
No, it would conflict with the internal loop() method that handles RAF checks. But it shouldn't read as a verb, I will change to allowsLooping so it sounds more like a property.

Reply:
Realized the rest of degu uses onRaf instead of loop. Renamed loop, which freed up the property name.